xds: Add a KeyValueStore-backed xDS delegate extension to contrib

abeyad opened the pull request:
An xDS delegate extension point was added in
https://github.com/envoyproxy/envoy/pull/22473 to enable custom behavior upon
receiving and loading xDS resources. This change creates an implementation of
the XdsResourcesDelegate interface that is backed by a KeyValueStore.

The intended use case is to enable persisting xDS resources and loading them on
startup in Envoy Mobile, in the event that the xDS control plane is unreachable.

Signed-off-by: Ali Beyad <abeyad@google.com>

Risk Level: Low
Testing: Unit & Integration tests
Docs Changes: N/A (hidden experimental feature)
Release Notes: N/A
Platform Specific Features: N/A

abeyad commented:
/assign @alyssawilk @adisuissa

alyssawilk commented:
/wait on CI :-P

abeyad commented:
/retest

alyssawilk commented:
@adisuissa you game for first pass on this?

adisuissa commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -90,6 +90,15 @@ void GrpcMuxImpl::sendDiscoveryRequest(absl::string_view type_url) {
   }
 
+  if (starting_) {
+    // On the initialization of the gRPC mux, load the persisted config, if available. If the xDS
```
This seems like it could "waste" memory and cpu resources, especially if 1) the save data is stale, 2) the server is ready and can easily serve config.
Shouldn't this be part of a fall-back option?

adisuissa reviewed (commented):
Thanks!
One high-level comment, instead of adding the extension under the "xds" directory, it should probably be under the path name "config".

abeyad commented on contrib/xds/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,130 @@
+std::vector<envoy::service::discovery::v3::Resource>
+KeyValueStoreXdsDelegate::getResources(const XdsSourceId& source_id,
+                                       const std::vector<std::string>& resource_names) const {
+  if (resource_names.empty()) {
```
Done

abeyad commented:
> Thanks! One high-level comment, instead of adding the extension under the "xds" directory, it should probably be under the path name "config".

Done, thanks for the review!  Addressed the comments and pushed up some commits.  PTAL!

alyssawilk commented:
still letting Adi drive here but you'll want to sync past the merge conflicts when you have a few :-)

abeyad commented:
> still letting Adi drive here but you'll want to sync past the merge conflicts when you have a few :-)

PR has been updated with the latest from upstream/main

alyssawilk commented:
/retest

adisuissa commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -321,7 +380,17 @@ void GrpcMuxImpl::onEstablishmentFailure() {
     }
+    if (starting_) {
+      // On the initialization of the gRPC mux, if connection to the xDS server fails, load the
+      // persisted config, if available. The locally persisted config will be used until
+      // connectivity is established with the xDS server.
```
Can you add an integration tests that emulates the "connectivity is re-established" scenario?

abeyad replied:
Added a test for this

adisuissa commented on contrib/config/test/kv_store_xds_delegate_integration_test.cc:
```diff
@@ -0,0 +1,340 @@
+  FakeStreamPtr rtds_stream_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, KeyValueStoreXdsDelegateIntegrationTest,
```
This should also be tested against unified-SotW

abeyad replied:
This PR doesn't handle unified-SotW yet.  I wanted to get it working for regular Sotw and then do the same for unified-Sotw in a follow up PR.

adisuissa commented on contrib/config/test/kv_store_xds_delegate_integration_test.cc:
```diff
@@ -0,0 +1,340 @@
+  checkSecretExists(std::string(CLIENT_CERT_NAME), /*version_info=*/"1");
+  EXPECT_EQ("whatevs", getRuntimeKey("foo"));
+  EXPECT_EQ("yar", getRuntimeKey("bar"));
+  EXPECT_EQ("saz", getRuntimeKey("baz"));
```
I guess this is the indication that the data was loaded from the KV-store.
To enhance observability I suggest adding a counter whenever data is loaded from the KV-store (for each type).

abeyad replied:
I added some stats

alyssawilk commented on api/contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.proto:
```diff
@@ -0,0 +1,23 @@
+
+// [#extension: envoy.xds_delegates.kv_store]
+//
+// Configuration for a KeyValueStore-based XdsResourcesDelegate implementation.
```
Comment what it should do?  (update KV store with latest config, etc)

abeyad replied:
Done.

alyssawilk commented on contrib/config/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,127 @@
+using ::envoy::extensions::config::v3alpha::KeyValueStoreXdsDelegateConfig;
+
+// TODO(abeyad): Add a per-Envoy instance prefix to the storage key, to allow xDS resources to be
+// persisted per-running-Envoy when Envoy is embedded as a library (e.g. Envoy Mobile).
```
Hm, this is an interesting one.  I wonder if we can get away with a winner-takes-all solution here - I'm interested in @RyanTheOptimist thoughts

RyanTheOptimist replied:
Hm. I'm not quite sure what is meant by winner-takes-all. Is the thinking that multiple Envoy instances would share the same KV store and whoever gets there first owns it? Or something else?

abeyad replied:
Actually, I think we can ensure different Envoy engines can have different key prefixes to avoid collisions by setting different keys in https://github.com/envoyproxy/envoy-mobile/blob/main/library/common/extensions/key_value/platform/platform.proto#L11

alyssawilk replied:
Ok in that case do we need a per-instance prefix?

abeyad replied:
We don't, I pushed a commit removing this comment.

alyssawilk commented on contrib/config/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,127 @@
+KeyValueStoreXdsDelegate::getResources(const XdsSourceId& source_id,
+                                       const std::vector<std::string>& resource_names) const {
+  if (resource_names.empty() || (resource_names.size() == 1 && resource_names[0] == "*")) {
+    // Empty names or one entry with "*" means wildcard.
```
optional nit: I mildly dislike having two wildcard options.  Could we validate that the resource name not be empty?

abeyad replied:
Unfortunately, Envoy allows these two different ways to specify wildcard (empty resource names or one resource name with "*").  Or do you mean, when invoking getResources, if resource_names is empty, add one `*` resource name to the list before passing in?

alyssawilk replied:
no I didn't know this was prior art.  Don't mind me then :-P

adisuissa commented on source/common/config/grpc_mux_impl.h:
```diff
@@ -178,6 +185,7 @@ class GrpcMuxImpl : public GrpcMux,
   XdsResourcesDelegateOptRef xds_resources_delegate_;
   const std::string target_xds_authority_;
   bool first_stream_request_;
+  bool starting_{true};
```
nit: starting is too generic. Maybe something in the lines of previously fetched data, or something that will convey the meaning of this

abeyad replied:
Done.

adisuissa reviewed (commented):
Overall LGTM.

I wonder how this is impacting warmed objects.
For example, if there was a CDS request that returned some EDS-type clusters, and then a disconnect happened; would you expect the endpoints to be loaded from the KV-store?
I think in the current implementation (assuming ADS is used) this will not happen, but it would be better to validate using an integration test.

alyssawilk reviewed (commented):
Very cool!

Are proto format failures real?
Also I'd definitely check coverage for this one as I don't think CI does coverage checks for contrib

abeyad commented on contrib/config/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,127 @@
+std::vector<envoy::service::discovery::v3::Resource>
+KeyValueStoreXdsDelegate::getAllResources(const XdsSourceId& source_id) const {
+  std::vector<envoy::service::discovery::v3::Resource> resources;
+  // TODO(abeyad): This is slow as we are iterating over all entries in the KV store; the
```
I like that and was hoping to do something like this.  Only try solving the problem if/when it becomes an issue down the road.

abeyad replied:
I added a comment to this effect.

alyssawilk commented on api/contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.proto:
```diff
@@ -0,0 +1,26 @@
+// servers, enabling configuration to be persisted locally and used on startup in case connectivity
+// with the xDS management servers could not be established.
+//
+// [#not-implemented-hide:]
```
I don't think we need to hide this given it's in contrib

abeyad replied:
Removed

alyssawilk commented on api/contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.proto:
```diff
@@ -0,0 +1,26 @@
+// servers, enabling configuration to be persisted locally and used on startup in case connectivity
+// with the xDS management servers could not be established.
+//
+// [#not-implemented-hide:]
```
did we want to comment the size "limitations"?

abeyad replied:
Done

alyssawilk reviewed (commented):
LG modulo one nit!

adisuissa commented on contrib/config/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,139 @@
+    resources = getAllResources(source_id);
+  } else {
+    for (const std::string& resource_name : resource_names) {
+      if (auto existing_resource = xds_config_store_->get(constructKey(source_id, resource_name))) {
```
bit: const

adisuissa commented on contrib/config/test/kv_store_xds_delegate_integration_test.cc:
```diff
@@ -0,0 +1,367 @@
+
+std::string kvStoreDelegateConfig() {
+  const std::string filename = TestEnvironment::temporaryPath("xds_kv_store.txt");
+  ::unlink(filename.c_str());
```
Can this use [this unlink](https://github.com/envoyproxy/envoy/blob/main/envoy/api/os_sys_calls.h#L212)?

abeyad replied:
done, and for the one below

adisuissa commented on contrib/xds/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,130 @@
+      r.set_version(decoded_resource.version());
+      r.mutable_resource()->PackFrom(decoded_resource.resource());
+      if (decoded_resource.ttl()) {
+        r.mutable_ttl()->CopyFrom(Protobuf::util::TimeUtil::MillisecondsToDuration(
```
Can you add a test for this?
I think there's a high-level requirements question here. If a resource has TTL of 1 minute, and it is stored in the KV store with that value, what should happen if it is read the day after? Should it still be 1 minute, or should it be discarded.

abeyad replied:
I think it should be discarded (e.g. imagine cached secrets that have expired... we probably don't want to use the expired certs).  I'm open to hearing other ideas, cc @alyssawilk 

RE: a test, I'm actually adding support for setting TTLs on the KV store in a follow-up PR, so I was going to add a test then.  I wanted to keep this PR more manageable, hence adding the TTL support in a follow up PR.

abeyad commented:
The proto format failures have disappeared, with adding the allow-fully-qualified-name annotation introduced in https://github.com/envoyproxy/envoy/pull/23163

abeyad commented on contrib/config/source/kv_store_xds_delegate.cc:
```diff
@@ -0,0 +1,146 @@
+      const std::string resource_key = constructKey(source_id, resource_name);
+      if (const auto existing_resource = xds_config_store_->get(resource_key)) {
+        envoy::service::discovery::v3::Resource r;
+        if (!r.ParseFromString(std::string(*existing_resource))) {
```
@adisuissa I added a check here if the resource fails to parse, including maintaining stats about it and removing the resource from the KV store if it's no longer parseable.

abeyad replied:
I pushed a commit that shows via a test that we handle invalid protos (e.g. proto enum value no longer valid) appropriately.

alyssawilk reviewed (approved):
LGTM, deferring to Adi for merge

abeyad commented:
/retest

abeyad commented:
@adisuissa friendly ping; I pushed up a commit containing an integration test that shows we handle invalid proto enums properly.

adisuissa commented on source/common/config/grpc_mux_impl.cc:
```diff
@@ -107,6 +107,53 @@ void GrpcMuxImpl::sendDiscoveryRequest(absl::string_view type_url) {
+  END_TRY
+  catch (const EnvoyException& e) {
+    // TODO(abeyad): do something more than just logging the error and incrementing a counter?
+    control_plane_stats.xds_local_load_failed_.inc();
```
This can probably be a call to delegate signaling that the config was rejected.
This way the delegate can free resources, if needed.

abeyad replied:
Done

adisuissa reviewed (commented):
Thanks, minor comment, but looks good otherwise.

abeyad commented on envoy/config/grpc_mux.h:
```diff
@@ -19,6 +19,7 @@ using ScopedResume = std::unique_ptr<Cleanup>;
  */
 #define ALL_CONTROL_PLANE_STATS(COUNTER, GAUGE, TEXT_READOUT)                                      \
   COUNTER(rate_limit_enforced)                                                                     \
+  COUNTER(xds_local_load_failed)                                                                   \
```
Thanks for the idea, totally agree it's a much better way.  Done!

abeyad reviewed (commented):
PTAL

adisuissa reviewed (approved):
LGTM, thanks!

abeyad commented:
/assign @lizan

abeyad commented:
/retest

abeyad commented:
Thanks for the reviews, everyone!
